### PR TITLE
Adding batchnorm verification

### PR DESCRIPTION
**Pull request by AkshitaB**
Changes proposed in this pull request:

- `BatchNormVerification` added as a sanity check.
- Updated `ModelTestCase` to run the sanity checks for all allennlp models.

### CODE REVIEW

**Comment by AkshitaB** [comment]
The verification check found invalid layer combinations in our QANet models!

**Comment by dirkgr** [comment]
> The verification check found invalid layer combinations in our QANet models!

@AkshitaB, what exactly did it find?

**Comment by dirkgr** [comment]
Looks like it found a `BatchNorm` layer after a layer with bias. @matt-gardner, is that an oversight in QANet?

File: allennlp/sanity_checks/batch_norm_verification.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    normalization_layer = (
+        nn.BatchNorm1d,
+        nn.BatchNorm2d,
+        nn.BatchNorm3d,
+        nn.SyncBatchNorm,
+        nn.InstanceNorm1d,
+        nn.InstanceNorm2d,
+        nn.InstanceNorm3d,
+        nn.GroupNorm,
+        nn.LayerNorm,
```
Comment (dirkgr) [inline allennlp/sanity_checks/batch_norm_verification.py:26]: Given that it detects all of these, can we call it something else? `NormalizationVerification` or something?

File: allennlp/training/trainer.py [hunk 2]
```diff
@@ -177,6 +178,50 @@ def on_end(
...
+    ) -> None:
+        if not is_primary:
+            return None
+
+        # We destroy the hooks after the first batch, since we only want to
+        # perform this check once.
+        if epoch == 0 and batch_number == 1 and is_training:
+            self._verification.destroy_hooks()
+            detected_pairs = self._verification.collect_detections()
+            # TODO: Should we actual fail with an error instead?
```
Comment (dirkgr) [inline allennlp/training/trainer.py:217]: Yes, but we should give a way to not run this check.

File: allennlp/training/trainer.py [hunk 3]
```diff
@@ -177,6 +178,50 @@ def on_end(
...
+            if detected_pairs:
+                logger.warning(
+                    "The model failed the BatchNormVerification check. "
+                    f"It contains {len(detected_pairs)} invalid layer combination(s)."
+                )
```
Comment (dirkgr) [inline allennlp/training/trainer.py:222]: Can we have a more descriptive error message, so people working on models can immediately find where the problem is?

File: tests/training/trainer_test.py [hunk 4]
```diff
@@ -755,6 +757,31 @@ def test_trainer_can_log_learning_rates_tensorboard(self):
 
         trainer.train()
 
+    def test_sanity_check_callback(self):
+        class BiasBatchNormModel(Model):
+            def __init__(self, use_bias=True):
```
Comment (dirkgr) [inline tests/training/trainer_test.py:762]: This has the `use_bias` parameter, but it's always `True`. Did you mean to parameterize this test to test both cases?

Comment (dirkgr) [reply tests/training/trainer_test.py:762]: Wait, this class is here twice. Can you re-use it from the other file?

**Review by dirkgr — COMMENTED** [review]
This is very cool. I wish we could think of 10 more checks like it.

Can you think of a way to run this all the time by default, but give a way to turn it off sometimes, without blowing up the config? I can't think of a way to do this without plumbing another parameter into `GradientDescentTrainer.from_params`. But maybe that's worth it?

**Comment by matt-gardner** [comment]
> Looks like it found a `BatchNorm` layer after a layer with bias. @matt-gardner, is that an oversight in QANet?

Quite possibly; I don't know the answer to this question, as I'm not super familiar with the internals of QANet, or with the batchnorm issue being checked here.  I wasn't the one who implemented this.

File: allennlp/sanity_checks/batch_norm_verification.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    def collect_detections(self):
+        detected_pairs = []
+        for (name0, mod0), (name1, mod1) in zip(
+            self._module_sequence[:-1], self._module_sequence[1:]
+        ):
+            bias = getattr(mod0, "bias", None)
+            detected = (
+                isinstance(mod1, self.normalization_layer)
+                and mod1.training  # TODO: do we want/need this check?
```
Comment (dirkgr) [inline allennlp/sanity_checks/batch_norm_verification.py:56]: I think the whole thing only makes sense if we're training. When evaluating, you might want to run all kinds of crazy stuff, because you're making baselines of somebody else's code.

**Comment by dirkgr** [comment]
@AkshitaB, can we fix it in QANet and re-train? It trains pretty quickly.

**Comment by AkshitaB** [comment]
@dirkgr Can you take a look at this? This PR will have to be merged because the parameter for ignoring the sanity checks currently failing is also added as part of this PR.

File: allennlp/training/trainer.py [hunk 6]
```diff
@@ -195,6 +196,47 @@ def on_end(
 TrainerCallback.register("null")(TrainerCallback)
 
 
+@TrainerCallback.register("sanity-checks")
+class SanityCheckCallback(TrainerCallback):
+    """
+    Performs model sanity checks.
+    """
```
Comment (dirkgr) [inline allennlp/training/trainer.py:203]: It would be nice if somewhere in this class it explained that even though we have only one sanity check right now, this is the place to add additional ones.

File: allennlp/sanity_checks/normalization_bias_verification.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                isinstance(mod1, self.normalization_layers)
+                and mod1.training
+                and isinstance(bias, torch.Tensor)
+                and bias.requires_grad
+            )
+            if detected:
+                detected_pairs.append((name0, name1))
+        self._detected_pairs = detected_pairs
+        if detected_pairs:
+            logger.warning(self._verification_message())
```
Comment (dirkgr) [inline allennlp/sanity_checks/normalization_bias_verification.py:69]: I thought we wanted to throw an exception if this fails? Otherwise it will get lost in a sea of messages during training.

Comment (AkshitaB) [reply allennlp/sanity_checks/normalization_bias_verification.py:69]: We throw an exception in the trainer callback.
